java/java.lsp.server/vscode/src/launcher.ts

```diff
@@ -0,0 +1,111 @@
+            const dir = path.join(e.extensionPath, 'nb-java-lsp-server');
+            if (!fs.existsSync(dir)) {
+                continue;
+            }
+            const exists = fs.readdirSync(dir);
+            for (let clusterName of exists) {
+                let clusterPath = path.join(dir, clusterName);
+                let clusterModules = path.join(clusterPath, 'config', 'Modules');
+                if (!fs.existsSync(clusterModules)) {
+                    continue;
+                }
+                let perm = fs.statSync(clusterModules);
+                if (perm.isDirectory()) {
+                    clusters.push(clusterPath);
+                }
+            }
+        }
+        return clusters;
+    }
+    let clusters = findClusters();
+
+    const nbexec = path.join(context.extensionPath, 'nb-java-lsp-server', 'platform', 'lib', 'nbexec');
+    let nbexecPerm = fs.statSync(nbexec);
+    if (!nbexecPerm.isFile()) {
+        throw `Cannot execute ${nbexec}`;
+    }
+
+    const userDir = path.join(context.extensionPath, "nb-java-lsp-server-user-dir");
+    fs.mkdirSync(userDir, {recursive: true});
+    let userDirPerm = fs.statSync(userDir);
+    if (!userDirPerm.isDirectory()) {
+        throw `Cannot create ${userDir}`;
+    }
+
+    let clusterPath = clusters.join(path.delimiter);
+    let ideArgs: string[] = [
+        "--nosplash", "--nogui", "--branding", "vscapp",
+        "-J-Djava.awt.headless=true",
+        "-J-Dnetbeans.logger.console=true",
+
+        "-J--add-opens=java.base/java.net=ALL-UNNAMED",
+        "-J--add-opens=java.base/java.lang.ref=ALL-UNNAMED",
+        "-J--add-opens=java.base/java.lang=ALL-UNNAMED",
+        "-J--add-opens=java.base/java.security=ALL-UNNAMED",
+        "-J--add-opens=java.base/java.util=ALL-UNNAMED",
+        "-J--add-opens=java.desktop/javax.swing.plaf.basic=ALL-UNNAMED",
+        "-J--add-opens=java.desktop/javax.swing.text=ALL-UNNAMED",
+        "-J--add-opens=java.desktop/javax.swing=ALL-UNNAMED",
+        "-J--add-opens=java.desktop/java.awt=ALL-UNNAMED",
+        "-J--add-opens=java.desktop/java.awt.event=ALL-UNNAMED",
+        "-J--add-opens=java.prefs/java.util.prefs=ALL-UNNAMED",
+        "-J--add-opens=jdk.jshell/jdk.jshell=ALL-UNNAMED",
+        "-J--add-modules=jdk.jshell",
+        "-J--add-exports=java.desktop/sun.awt=ALL-UNNAMED",
+        "-J--add-exports=java.desktop/java.awt.peer=ALL-UNNAMED",
+        "-J--add-exports=java.desktop/com.sun.beans.editors=ALL-UNNAMED",
+        "-J--add-exports=java.desktop/sun.swing=ALL-UNNAMED",
+        "-J--add-exports=java.desktop/sun.awt.im=ALL-UNNAMED",
+        "-J--add-exports=jdk.internal.jvmstat/sun.jvmstat.monitor=ALL-UNNAMED",
+        "-J--add-exports=java.management/sun.management=ALL-UNNAMED",
+        "-J--add-exports=java.base/sun.reflect.annotation=ALL-UNNAMED",
+        "-J-XX:+IgnoreUnrecognizedVMOptions",
+
+        "--clusters", clusterPath, "--userdir", userDir
```

Comment:
what about `--cachedir`

Reply:
It isn't specified in current `etc/nb-java-lsp-server.conf` file either. As such it defaults to `$userdir/var/cache`. By placing `userDir` under the extension directory, this change also places the cache directory there.

I know the NetBeans IDE is using `$HOME/.cache/netbeans/12.x/`. Shall I try to use such location as well? I cannot find a OS neutral API for that in node.js documentation...

Reply:
IMHO depends on where vscode extensions typically store their cache/temp files (?)

Reply:
OK, I'll modify the code to use [globalStoragePath](https://github.com/microsoft/vscode/blob/1fd7ee15d3a2c50547228b08cefc7c13a5a739d2/src/vs/vscode.d.ts#L5716) then. That's the only [suggestion I could find](https://github.com/microsoft/vscode/issues/14476).